posts/2025-11-13-langchain4j-jakartaee-streaming.adoc

```diff
@@ -0,0 +1,212 @@
+
+== Try out the streaming chat sample application
+
+You can check out a sample application that integrates LangChain4j's streaming chat with a Jakarta EE application. Clone the https://github.com/OpenLiberty/sample-langchain4j.git[sample-langchain4j] repository and go into the streaming-chat project by using the following commands:
+
+[source,console]
+----
+$ git clone https://github.com/OpenLiberty/sample-langchain4j.git
+$ cd sample-langchain4j/streaming-chat
+----
+
+Install Java 21 or newer if you haven’t already. You can use https://developer.ibm.com/languages/java/semeru-runtimes/downloads[IBM Semeru Runtimes] as your chosen Java runtime. Set the `JAVA_HOME` environment variable to your Java installation path.
+
+The application supports the GitHub, Ollama, and Mistral AI providers. Set up an AI provider by following the https://github.com/OpenLiberty/sample-langchain4j/blob/main/streaming-chat/README.md[Prerequisites and Environment Set Up instructions].
+
+Start the sample application by running `./mvnw liberty:dev` if you are on MacOS or Linux, and `mvnw.cmd liberty:dev` if you are on Windows. When the server starts, go to the http://localhost:9080 URL, send a question (for example "What are large language models?"), and check out how the response is streamed token by token.
+
+== How does the application work?
+
+From a high level, the user connects from their front end to a WebSocket endpoint. Each user message is passed to the AI service, which sends a request to the LLM provider. As the response is generated, the WebSocket endpoint sends each token to the user, rather than waiting until the entire response is generated.
+
+image::/img/blog/langchain4jStreamingDiagram.png[Architectural diagram,width=70%,height=70%,align="center"]
+
+The application builds a streaming chat model from a provider of your choice and sets a 200 token limit on the response. For example, the following code builds a streaming chat model that uses Ollama's Llama3.2 model.
+
+[source,java]
+----
+StreamingChatModel streamingModel = OllamaStreamingChatModel.builder()
+    .baseUrl("http://localhost:11434")
+    .modelName("llama3.2")
+    .numPredict(200)
+    .build();
+----
+
+See the https://github.com/OpenLiberty/sample-langchain4j/blob/7112a4f14bc7b63bd97f8f123f6ba3c16fdebba9/streaming-chat/src/main/java/io/openliberty/sample/langchain4j/util/ModelBuilder.java#L93-L99[ModelBuilder.java file] to learn more.
+
+An assistant interface with a chat method is declared for LangChain4j to implement. To support multiple user conversations at the same time, the ChatMemoryAccess interface is extended to provide a method that deletes conversation history such as when a user exits. You can also specify a system message to influence the assistant's responses, such as being knowledgeable about the Open Liberty application server.
```

Comment:
I'm not entirely sure how deleting conversation history is a way of supporting multiple user conversations at the same time?